### PR TITLE
Add Jest linting and enforce test consistency

### PR DESCRIPTION
**Pull request by MrTimscampi**
Upstreaming of my changes in regards to tests.

* Adds Jest linting (Enforces best practices and use of `it()` inside `describe()` blocks)
* Rework and document the message format
* Add threshold for Codecov patch job and don't fail if coverage is not found (Taken from [Google's Lighthouse](https://github.com/GoogleChrome/lighthouse/blob/master/.codecov.yml))
* Fixes an issue with setPreviousTrack not updating the lastItemIndex
* Adds a bunch of TODOs mentioning refactors needed in tests for proper unit testing

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -35,8 +35,49 @@ The following should guide you through making a proper commit message:
...
-## Code style
+### Code style
 
 This repository uses [Prettier](https://prettier.io/) for automatically formating code. In addition, [ESLint](https://eslint.org/) rules are provided to help you enforce good practices.
 
 The project is written in [TypeScript](typescriptlang.org/), and using strongly typed code is recommended and expected, where possible.
+
+### Tests
+
+We use [Jest](https://jestjs.io/) for unit testing. For testing Vue components, usage of [vue-testing-library](https://testing-library.com/docs/vue-testing-library/intro/) is prefered.
```
Comment (ferferga) [inline CONTRIBUTING.md:46]:
```suggestion
We use [Jest](https://jestjs.io/) for unit testing. For testing Vue components, usage of [vue-testing-library](https://testing-library.com/docs/vue-testing-library/intro/) is preferred.
```

**Review by ferferga — CHANGES_REQUESTED** [review]

File: store/__tests__/backdrop.spec.ts [hunk 2]
```diff
@@ -31,70 +31,76 @@ beforeEach(() => {
...
-  store.replaceState({ ...BACKDROP_SET_TEST_VALUE });
+  it('clears the hash when "CLEAR_CURRENT_BACKDROP" is committed', () => {
+    store.replaceState({ ...BACKDROP_SET_TEST_VALUE });
 
-  store.commit('CLEAR_CURRENT_BACKDROP');
+    store.commit('CLEAR_CURRENT_BACKDROP');
 
-  expect(store.state.blurhash).toBe('');
-});
+    expect(store.state.blurhash).toBe('');
```
Comment (ThibaultNocchi) [inline store/__tests__/backdrop.spec.ts:56]: Can't we replace the `''` by the value from defaultState or something?

File: store/__tests__/backdrop.spec.ts [hunk 3]
```diff
@@ -31,70 +31,76 @@ beforeEach(() => {
...
-  store.replaceState({ ...BACKDROP_SET_TEST_VALUE });
+  it('resets the opacity when "RESET_BACKDROP_OPACITY" is committed', () => {
+    store.replaceState({ ...BACKDROP_SET_TEST_VALUE });
 
-  store.commit('RESET_BACKDROP_OPACITY');
+    store.commit('RESET_BACKDROP_OPACITY');
 
-  expect(store.state.opacity).toBe(0.75);
-});
+    expect(store.state.opacity).toBe(0.75);
```
Comment (ThibaultNocchi) [inline store/__tests__/backdrop.spec.ts:64]: Same here

File: store/__tests__/backdrop.spec.ts [hunk 4]
```diff
@@ -31,70 +31,76 @@ beforeEach(() => {
...
+  it('clears the hash when clearBackdrop is dispatched', () => {
+    // TODO: This should only test if the proper mutation is committed
+    store.replaceState({ ...BACKDROP_SET_TEST_VALUE });
 
-  store.dispatch('clearBackdrop');
+    store.dispatch('clearBackdrop');
 
-  expect(store.state.blurhash).toBe('');
-});
+    expect(store.state.blurhash).toBe('');
```
Comment (ThibaultNocchi) [inline store/__tests__/backdrop.spec.ts:84]: Same here

File: store/__tests__/userViews.spec.ts [hunk 5]
```diff
@@ -40,60 +40,62 @@ beforeEach(() => {
...
+    store.replaceState({
+      views: [DEMO_TEST_ITEM_A, DEMO_TEST_ITEM_B, DEMO_TEST_ITEM_C]
+    });
 
-test('When "CLEAR_USER_VIEWS" is committed, user views are cleared.', () => {
-  store.replaceState({
-    views: [DEMO_TEST_ITEM_A, DEMO_TEST_ITEM_B, DEMO_TEST_ITEM_C]
+    store.commit('CLEAR_USER_VIEWS');
+
+    expect(store.state.views).toMatchObject([]);
```
Comment (ThibaultNocchi) [inline store/__tests__/userViews.spec.ts:65]: We can use the defaultstate value

File: store/__tests__/userViews.spec.ts [hunk 6]
```diff
@@ -40,60 +40,62 @@ beforeEach(() => {
...
+    });
 
-  expect(store.state.views).toMatchObject([]);
-});
+    store.dispatch('clearUserViews');
 
-test('When clearUserViews is is called, user views are cleared.', () => {
-  store.replaceState({
-    views: [DEMO_TEST_ITEM_A, DEMO_TEST_ITEM_B, DEMO_TEST_ITEM_C]
+    expect(store.state.views).toMatchObject([]);
```
Comment (ThibaultNocchi) [inline store/__tests__/userViews.spec.ts:75]: Use defaultstate

File: store/__tests__/userViews.spec.ts [hunk 7]
```diff
@@ -40,60 +40,62 @@ beforeEach(() => {
...
+    ]);
   });
 
-  expect(store.getters.getNavigationDrawerItems).toMatchObject([
-    { icon: 'mdi-movie', title: 'test-view-a', to: '/library/test-id-1' },
-    { icon: 'mdi-music', title: 'test-view-b', to: '/library/test-id-2' },
-    { icon: 'mdi-folder', title: 'test-view-c', to: '/library/test-id-3' }
-  ]);
-});
+  it('gets all the user views when getUserViews is callsed dispatched', () => {
```
Comment (ThibaultNocchi) [inline store/__tests__/userViews.spec.ts:90]: Typo and weird sentence :p

File: store/__tests__/tvShows.spec.ts [hunk 8]
```diff
@@ -46,76 +46,78 @@ beforeEach(() => {
...
+    store.commit('CLEAR_TVSHOWS');
 
-  expect(store.state).toMatchObject(defaultState());
-});
+    expect(store.state).toMatchObject(defaultState());
+  });
 
-test('When getTvShowsSeasonEpisodesSuccess is called, showNavDrawer is set.', () => {
-  store.replaceState({ ...defaultState() });
+  it('sets the show, seasons and episodes <hen getTvShowsSeasonEpisodesSuccess is dispatched', () => {
```
Comment (ThibaultNocchi) [inline store/__tests__/tvShows.spec.ts:93]: Typo

File: store/__tests__/snackbar.spec.ts [hunk 9]
```diff
@@ -25,49 +25,51 @@ beforeEach(() => {
   store = new Vuex.Store(cloneDeep({ state, mutations, actions }));
 });
 
-test('When "SET_SNACKBAR_MESSAGE" is committed, the snackbar message is set.', () => {
-  store.replaceState({ ...defaultState() });
+describe('vuex: snackbar', () => {
+  it('sets the snackbard message when "SET_SNACKBAR_MESSAGE" is committed', () => {
```
Comment (ThibaultNocchi) [inline store/__tests__/snackbar.spec.ts:29]: Snackbard lmao, I'm thinking of a snackbar with a beard and some luth or something

File: store/__tests__/snackbar.spec.ts [hunk 10]
```diff
@@ -25,49 +25,51 @@ beforeEach(() => {
...
+    store.commit('SET_SNACKBAR_MESSAGE', SNACKBAR_SET_TEST_VALUE);
 
-  expect(store.state).toMatchObject(SNACKBAR_SET_TEST_VALUE);
-});
+    expect(store.state).toMatchObject(SNACKBAR_SET_TEST_VALUE);
+  });
 
-test('When "RESET_MESSAGE" is committed, the message is reset.', () => {
-  store.replaceState({ ...SNACKBAR_SET_TEST_VALUE });
+  it('resets the snackbard message when "RESET_MESSAGE" is committed', () => {
```
Comment (ThibaultNocchi) [inline store/__tests__/snackbar.spec.ts:37]: Snackbard's back

File: store/__tests__/snackbar.spec.ts [hunk 11]
```diff
@@ -25,49 +25,51 @@ beforeEach(() => {
...
-  expect(store.state).toMatchObject(SNACKBAR_SET_TEST_VALUE);
-});
+    expect(store.state).toMatchObject(SNACKBAR_SET_TEST_VALUE);
+  });
 
-// Undefined color is '' when color is not passed
-test('When pushSnackbarMessage is called, snackbar message is set. Case B', () => {
-  store.replaceState({ ...defaultState() });
+  // Undefined color is '' when color is not passed
+  it('sets the snackback message and no color when pushSnackbarMessage is dispatched without a color', () => {
```
Comment (ThibaultNocchi) [inline store/__tests__/snackbar.spec.ts:54]: Snackback ;p

File: store/__tests__/playbackManager.spec.ts [hunk 12]
```diff
@@ -52,721 +52,745 @@ beforeEach(() => {
...
 
-  expect(store.state.lastItemIndex).toBe(3);
-  expect(store.state.currentItemIndex).toBe(5);
-});
+    store.commit('CLEAR_QUEUE');
 
-test('When "SET_CURRENT_MEDIA_SOURCE" is committed, queue is cleared.', () => {
-  store.replaceState({
-    ...defaultState()
+    expect(store.state.queue).toMatchObject([]);
```
Comment (ThibaultNocchi) [inline store/__tests__/playbackManager.spec.ts:237]: Use defaultstate :)

File: store/__tests__/playbackManager.spec.ts [hunk 13]
```diff
@@ -52,721 +52,745 @@ beforeEach(() => {
...
 
-  expect(store.state.currentTime).toBe(0);
-});
+    store.dispatch('resetCurrentTime');
 
-test('When resetCurrentItemIndex is is called, the currentItemIndex is set to null.', () => {
-  store.replaceState({
-    ...defaultState(),
-    currentItemIndex: 1
+    expect(store.state.currentTime).toBe(0);
```
Comment (ThibaultNocchi) [inline store/__tests__/playbackManager.spec.ts:583]: Use defaultstate?

File: store/__tests__/playbackManager.spec.ts [hunk 14]
```diff
@@ -52,721 +52,745 @@ beforeEach(() => {
...
 
-  expect(store.state.currentTime).toBe(0);
-});
+    store.dispatch('skipBackward');
 
-test('When skipBackward is is called, the currentTime is updated. Case B', () => {
-  store.replaceState({
-    ...defaultState(),
-    currentTime: 25
+    expect(store.state.currentTime).toBe(0);
```
Comment (ThibaultNocchi) [inline store/__tests__/playbackManager.spec.ts:701]: Defaultstate ?

**Review by ThibaultNocchi — COMMENTED** [review]
A few nitpicks

File: store/__tests__/playbackManager.spec.ts (hunk 13 above)
Comment (MrTimscampi) [reply store/__tests__/playbackManager.spec.ts:594]: It has to be 0, the default state is null.

File: store/__tests__/playbackManager.spec.ts (hunk 14 above)
Comment (MrTimscampi) [reply store/__tests__/playbackManager.spec.ts:723]: It has to be 0, the default state is null.

**Review by ThibaultNocchi — APPROVED** [review]
LGTM

**Review by ferferga — CHANGES_REQUESTED** [review]
Why all the ``@ts-expect-error`` were replaced by ``@ts-ignore``?

**Comment by MrTimscampi** [comment]
> Why all the `@ts-expect-error` were replaced by `@ts-ignore`?

Because git rebase sucks :p Give me a minute

File: components/Layout/Images/__tests__/BlurhashCanvas.spec.ts [hunk 15]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable @typescript-eslint/ban-ts-comment */
```
Comment (ferferga) [inline components/Layout/Images/__tests__/BlurhashCanvas.spec.ts:1]: You can remove this and add the descriptions that were alongside ``@ts-expect-error`` comments so ESLint doesn''t complain

**Review by ferferga — CHANGES_REQUESTED** [review]

**Review by ferferga — APPROVED** [review]

**Pull request merged**